### PR TITLE
Add script to load Candid data

### PR DESCRIPTION
**Pull request by jasonaowen**
This is by no means finished, but I wanted to get something out before I left in case y'all are interested in looking before I get back to this.

### CODE REVIEW

**Comment by jasonaowen** [comment]
@slifty this should be far enough along for you to get some sense of what's involved on the Charity Navigator side; I'd also be happy to split the first few commits out into their own PR if there's enough there for you to start building without wanting to wait on this one.

File: README.md [hunk 1]
```diff
@@ -2,14 +2,41 @@
...
 
 `npm run [script name] -- --[parameter name]=[arg]`
 
 To find information on one of the scripts, look in the script itself at `src/[script name].ts`.
+
+## Unified scripts
+
+Additional scripts are created as subcommands of a single, unified script.
+Run it with `npm start` to see the list of commands and required options.
+Pass arguments with a `--` between the name and its arguments: `npm start -- help`.
```
Comment (bickelj) [inline README.md:20]: I get three different results for `npm run start --help`, `npm start --help`, and `npm start -- help`, the last of which is the weirdest. I have a hard time recalling whether to run `npm run start` or `npm start` and what the difference is. So I don't know if a better command is available. But regardless I like the approach.

File: src/index.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { config } from 'dotenv';
 import yargs from 'yargs/yargs';
 import { hideBin } from 'yargs/helpers';
+import { logger } from './logger';
```
Comment (bickelj) [inline src/index.ts:5]: I think having a way to filter logs by class/module/source-file is essential. Is there an easy way to do that with winston?

File: README.md (hunk 1 above)
Comment (jasonaowen) [reply README.md:20]: > I get three different results for `npm run start --help`, `npm start --help`, and `npm start -- help`, the last of which is the weirdest. 

The first is giving you help for `npm run`, as it is ignoring `start` as a value for `<command>`. The second is giving you help for `npm start`, which is the preferred way to run a package in a dev environment. The third is passing the `help` argument to the script added in this PR, and can also be written `npm start -- --help` - both are auto-generated by yargs.

> I have a hard time recalling whether to run `npm run start` or `npm start` and what the difference is. So I don't know if a better command is available.

The `start`, `stop`, `restart`, and `test` commands are shortcuts to common scripts. `npm start` and `npm run start` are effectively the same, and either can be used here - but both require the `--` argument to pass subsequent arguments to the script itself instead of to npm.

> But regardless I like the approach.

Thanks! :blue_heart:

Comment (slifty) [reply README.md:20]: `npm start` is a special npm command that invokes whatever the defined `start` script is according to `package.json`
`npm run {NAME_OF_SCRIPT}` will run the script defined in `package.json` with `NAME_OF_SCRIPT`

`npm {command} {-flag}` will pass the flag / parameter values to `npm` itself
`npm {command} -- {-flag}` will pass the flag / parameter values after `--` to the called script

(hopefully I didn't make any mistakes)

File: package.json [hunk 3]
```diff
@@ -27,6 +29,11 @@
   "dependencies": {
     "axios": "^1.4.0",
     "csv-reader": "^1.0.12",
-    "ts-command-line-args": "^2.5.0"
+    "dotenv": "^16.0.3",
+    "jwt-decode": "^3.1.2",
+    "openid-client": "^5.4.2",
+    "ts-command-line-args": "^2.5.0",
+    "winston": "^3.8.2",
```
Comment (slifty) [inline package.json:36]: We're using pino in the other PDC project -- any particular reason not to normalize on that?

File: src/index.ts (hunk 2 above)
Comment (jasonaowen) [reply src/index.ts:5]: Say more? Is that filtering something you expect/want to do in this repo, or a capability you think is important in logging systems in general? How is that done in pino?

Does having [multiple loggers](https://github.com/winstonjs/winston/blob/master/README.md#working-with-multiple-loggers-in-winston) address that concern?

Comment (slifty) [reply src/index.ts:5]: In pino I believe it's done by providing a `source` -- see https://github.com/PhilanthropyDataCommons/service/blob/b04d4178a32fc5a03ffcb56acf066d4c23021549/src/logger.ts#L28

File: package.json (hunk 3 above)
Comment (jasonaowen) [reply package.json:36]: There was a [brief discussion on Zulip](https://chat.opentechstrategies.com/#narrow/stream/75-PDC/topic/Logging.20in.20node/near/165086), but broadly, `pino-pretty` is an inconvenient extra step - especially around argument passing - and Winston offers first class pretty formatting support and human readable timestamps. I wouldn't want to use locale time strings in prod, but for dev-facing-only scripts, I find them to be more readable.

File: src/pdc-api.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  callPdcApi<ApiBaseField[]>(
+    baseUrl,
+    '/baseFields',
+    {},
+    token,
+    'get',
+  )
+);
+
+interface ApiProposal {
```
Comment (bickelj) [inline src/pdc-api.ts:44]: Having re-written these interfaces in some of the other scripts, I think we might be at the point of benefiting from publishing a `@types/pdc` package from the `service` repo and importing those types here.

File: src/index.ts (hunk 2 above)
Comment (jasonaowen) [reply src/index.ts:5]: Ah, that's helpful - so yes I think we could easily do the same here if we want.

How is the filtering done, though - like how is that capability applied in practice?

File: package.json (hunk 3 above)
Comment (slifty) [reply package.json:36]: Ahh got it I missed it -- I do believe  pino pretty / timestamp formats can be enabled as part of the config [pretty config](https://github.com/pinojs/pino/blob/HEAD/docs/pretty.md).

Timestamp is just using a [customPrettifier](https://github.com/pinojs/pino-pretty#options) that would convert to whatever timestamp you want.

Not asking you to rewrite using pino or anything, but wanna just be sure it's clear that pino I believe is pretty much feature complete here and there is *some* cost in adding a new logging package that diverges from the project's existing one, since now the team has to become familiar with both.

For context I had initially used winston on the service repo long long ago and we chose pino after some discussion.

Comment (jasonaowen) [reply package.json:36]: Aha, I didn't realize it could be done in-app! That's a strong argument.

> Not asking you to rewrite using pino or anything

Hopefully that wouldn't be too big a lift; you're welcome to take a pass on that while I'm working on wrapping up Candid / testing https://github.com/PhilanthropyDataCommons/service/pull/373 !

> For context I had initially used winston on the service repo long long ago and we chose pino after some discussion.

Oh, interesting! Do you have a link to that discussion? What were the tradeoffs?

File: src/index.ts (hunk 2 above)
Comment (jasonaowen) [reply src/index.ts:5]: @bickelj I largely consider that to be noise outside of exceptions :upside_down_face: it isn't currently present, no.

File: package.json (hunk 3 above)
Comment (slifty) [reply package.json:36]: I LIED!  I didn't do the initial logging at all and was just merging OTS projects in my head.  But anyway here's the original issue / related commit logs https://github.com/PhilanthropyDataCommons/service/issues/41

Comment (bickelj) [reply package.json:36]: I'm OK with the timestamp being RFC-ized by default. The reason for not doing it out of the box is it breaks pino's performance claims. I think having a single JSON line is a feature whereby `grep` works to find my exact log statements of interest. If it's broken into multiple lines by default, now I have add `-C 100` or whatever. Regardless, there may be newlines within a single message, so this point may be moot.

Comment (jasonaowen) [reply package.json:36]: Right, that format makes sense in prod! I don't think it fits well in this repo, though.

All that said, at some point I think we're going to want a more sophisticated logging platform than syslog and grep. We don't need to have that conversation now, though.

**Comment by jasonaowen** [comment]
Okay, I pushed the scripts that I used to test https://github.com/PhilanthropyDataCommons/service/pull/373 . I'm still not done; in particular, I still need to handle rate limits (by sleeping), upload in the bulk case, resolve linter errors, and rebase to address merge conflicts from recent dependabots.

It sounds like both @slifty and @bickelj are against using winston for logging here, so setting up pino with a useful configuration is on my list, too. If either of you want to take a pass at that, I'd welcome the help.

In the meantime, I'm going to go eat!

**Comment by jasonaowen** [comment]
Still have some cleanup to do, but it *should* now successfully upload Candid data for each proposal in the PDC. Hard to test for sure without fully importing the database locally, which I don't want to do.

I am beginning to suspect that, in the oh-so-grand tradition of this repo, we will run the script and use the results before the PR introducing the script is merged.

File: src/candid.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    )
+    .command(
+      'update <ein>',
+      'Fetch information about an organization by its EIN, and upload to the PDC',
+      oidcOptions,
+      async (args) => {
+        const token = await getToken(
+          args.oidcBaseUrl as string,
+          args.oidcClientId as string,
+          args.oidcClientSecret as string,
```
Comment (jasonaowen) [inline src/candid.ts:109]: I'm not a fan of this, nor of the sequence of type guards alternative, but either the types for yargs or TypeScript itself doesn't allow for statically typing these slightly-dynamic options. If you have any suggestions for best to tell TypeScript that "hey, this can't be `undefined`, actually", I'd love to hear them!

File: src/candid.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      'update-all',
+      'Update Candid profiles for all PDC proposals',
+      oidcOptions,
+      async (args) => {
+        const token = await getToken(
+          args.oidcBaseUrl as string,
+          args.oidcClientId as string,
+          args.oidcClientSecret as string,
+        );
+        const eins = await getEinsFromPdc(args.pdcApiBaseUrl as string, token);
```
Comment (jasonaowen) [inline src/candid.ts:132]: I'm also not a fan of this pattern of passing a token + the base URL around. An alternative I considered is to make a PDC client class that initializes an Axios client, keeps it in the instance, and has a bunch of methods that use it; another alternative is to pass that `AxiosInstance` around directly. I'm kinda leaning towards the latter, but haven't implemented it yet.

File: src/candid.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        logger.info(`Found ${eins.length} valid EINs to look up in Candid`);
+
+        for (let i = 0; i < eins.length; i += 1) {
+          /* eslint-disable no-await-in-loop -- rate limit ourselves so Candid doesn't */
```
Comment (jasonaowen) [inline src/candid.ts:136]: If you have a better option than an old-school `for` loop that won't fire off every request at the same time, please share!

File: src/pdc-api.ts (hunk 4 above)
Comment (slifty) [reply src/pdc-api.ts:43]: We might at some point actually consider releasing a full blown SDK / client for the service API (which would include types)

Comment (jasonaowen) [reply src/pdc-api.ts:43]: Does OpenAPI / Swagger help with this? In theory it should be able to automatically generate such SDKs for multiple languages, but I haven't actually looked in to using that in practice.

File: src/candid.ts (hunk 7 above)
Comment (slifty) [reply src/candid.ts:136]: I do!  I bet you're not gonna like it though.

The only way I've found to have a series of promises that wait on one another is to use an async reduce.  Here's an example: https://github.com/PermanentOrg/sftp-service/blob/a1a0d740c7100e6689ead2953ddc41228db9f142/src/classes/PermanentFileSystem.ts#L346-L359

I believe that a for with an await can, in some cases, result in blocking code / starve the event loop, so I would recommend avoiding it (though since you're awaiting on I/O here I think it ends up being OK).

Separately from all that though, what do you think of updating `getCandidProfile` to `getCandidProfiles` and having it take an array / be responsible for rate limiting and negotiating with chunking against the API.

For an example of why I think that breakdown would be a cleaner separation of responsibilities: in Charity Navigator the `get` endpoint they provide allows you to send multiple EINs so I will be having `getCharityNavigatorProfiles` that encapsulates the negotiation with the API appropriately.

Comment (jasonaowen) [reply src/candid.ts:136]: > The only way I've found to have a series of promises that wait on one another is to use an async reduce.

Hah, I saw that while I was exploring my options. Do you think that's *better*? I ask sincerely, because I don't know if I have an opinion one way or another - as we've discussed, I tend to find `reduce`rs hard to read / reason about, but having to silence eslint warnings is also Not Great. 

> I believe that a for with an await can, in some cases, result in blocking code / starve the event loop, so I would recommend avoiding it (though since you're awaiting on I/O here I think it ends up being OK).

Can you say more? I'm not sure I follow!

> Separately from all that though, what do you think of updating `getCandidProfile` to `getCandidProfiles` and having it take an array / be responsible for rate limiting and negotiating with chunking against the API.

Sure, that sounds like a good idea to me!

Comment (slifty) [reply src/candid.ts:136]: > Do you think that's better?

I do prefer the use of reduce to a for loop, but I appreciate that reduce is more complex.  What I *really* wish is that something like Promise.sequence() existed and then you could just map the elements to a list of promises.

You probably saw this but holy moly (I think I prefer reduce to this, but hey, it's an option):

```
function runSerial(tasks) {
  var result = Promise.resolve();
  tasks.forEach(task => {
    result = result.then(() => task());
  });
  return result;
}
```
from https://stackoverflow.com/a/31070150/159522

> Can you say more? I'm not sure I follow!

If those links (specifically the first one / https://snyk.io/blog/nodejs-how-even-quick-async-functions-can-block-the-event-loop-starve-io/) don't explain it well enough then I'd probably be better of explaining by voice.  It involves some pretty low level base knowledge of the node event loop (which you may have! but I'd rather navigate that by voice).

Basically using await in a for loop does NOT mean that your loop is necessarily non-blocking; you can have a situation where microtasks generate more microtasks (and the way the event loop works, the loop doesn't continue until that queue is completely depleted).  I don't believe that applies here because of the use of I/O in the await, but by disabling the eslint rule we open the doors for unwittingly allowing it.

Comment (jasonaowen) [reply src/candid.ts:136]: > Basically using await in a for loop does NOT mean that your loop is necessarily non-blocking; you can have a situation where microtasks generate more microtasks (and the way the event loop works, the loop doesn't continue until that queue is completely depleted). I don't believe that applies here because of the use of I/O in the await, but by disabling the eslint rule we open the doors for unwittingly allowing it.

Ah, that is informative, thank you!

To expand on what you said, not only are we doing I/O rather than heavy computations, I think we also *don't care about the event loop* in this context - right? This is a single-purpose batch script being run until completion, so I don't think there are any events we would care about - with the possible exception (ha) of system signals, like `SIGINT` (often generated by `Ctrl-C`), which I don't think go through the event loop, but I am not sure about that.

All of that said, this situation is specifically called out in the [documentation for the `no-await-in-loop` rule](https://eslint.org/docs/latest/rules/no-await-in-loop):

> Or, loops may be used to prevent your code from sending an excessive amount of requests in parallel. In such cases it makes sense to use `await` within a loop and it is recommended to disable the rule via a standard ESLint disable comment.

So maybe I should get over my own distaste of `eslint-disable` comments?

>  What I really wish is that something like `Promise.sequence()` existed and then you could just map the elements to a list of promises.

Oh, that's a good call - is there a library we could lean on for that? I see a few options on npm, but none of them are very recent; I haven't dug in yet to see if that's because they're mature or abandoned.

Comment (slifty) [reply src/candid.ts:136]: > don't care about the event loop in this context

Iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii get what you're saying in theory but it's hard for me to imagine feeling OK with code that locks out I/O (we *do* care about the event loop since, for instance, network events are events too and we need to be able to consume network responses!) But yeah it's a moot point since we're not doing it.  Blocking the event loop means node isn't able to revisit async code blocks, for instance.  Idk!

Basically "don't block the event loop" is one of very few top level guides published by node: https://nodejs.org/en/docs/guides/dont-block-the-event-loop

> this situation is specifically called out

Hah the explicit callout is pretty strong evidence that it's a totally fine way to go.  (bottom line is that in a vaccum I'd use reduce, hence the example I linked to, but I think the foreach is fine)

I +1 the distaste to lint rule avoidance, but being able to say in a comment "this is explicitly suggested by the rule itself" is as iron clad as one can get.

Comment (slifty) [reply src/candid.ts:136]: ( oh and yeah I couldn't find any libraries for it last time I looked :( -- I think folks were like "just... use reduce!")

File: src/pdc-api.ts (hunk 4 above)
Comment (slifty) [reply src/pdc-api.ts:43]: Oooooh that'd be awesome.  Here are some resources related to that:

- https://github.com/OpenAPITools/openapi-generator
- https://github.com/swagger-api/swagger-codegen

Worth exploring, and probably good enough for our context.  More broadly though there are a few different ways to implement an SDK (e.g. functional vs object oriented) so I'm pretty curious if the quality / tree shaking potential of the resulting SDKs are comparable to a home brewed one!

File: src/candid.ts (hunk 7 above)
Comment (jasonaowen) [reply src/candid.ts:136]: So @slifty did we land on "this is fine, actually"? Or would you still prefer to use reduce?

**Comment by jasonaowen** [comment]
Now that https://github.com/PhilanthropyDataCommons/service/pull/373 has been deployed - thank you, @bickelj - it turns out we have some EINs that Candid does not know about, so I still need to handle 404s.

@slifty, how're you doing with Charity Navigator? Do you want to just add on to this PR? Or what are you thinking? Sorry this has been such a mess :/

File: src/candid.ts (hunk 7 above)
Comment (slifty) [reply src/candid.ts:136]: I think this is fine!

**Comment by slifty** [comment]
@jasonaowen I think it'd be cleaner to have Charity Navigator as a separate PR (I have it on another branch from this one / am fine rebasing as needed).  I was with family today for the holiday, but should have Charity Navigator data in the API tomorrow.

File: src/candid.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+          const ein = eins[i];
+          if (typeof ein !== 'string') {
+            break;
+          }
+          const data = await getCandidProfile(args.candidApiKey as string, ein);
+          await postPlatformProviderData(
+            args.pdcApiBaseUrl as string,
+            token,
+            ein,
+            'Candid',
```
Comment (jasonaowen) [inline src/candid.ts:146]: It turns out the front end will be simpler if we make these URL-convenient values. @slifty, can you please use `charity-navigator` when you post your data?

```suggestion
            'candid',
```

**Comment by jasonaowen** [comment]
> @jasonaowen I think it'd be cleaner to have Charity Navigator as a separate PR (I have it on another branch from this one / am fine rebasing as needed).

Okay. I'm deep into the front-end right now, and probably won't have a chance to work on this for a few days. Again, I'm sorry to open this big sprawling PR - particularly with the logging misstep - and then basically have no bandwidth to work on it further!

> I was with family today for the holiday, but should have Charity Navigator data in the API tomorrow.

I won't be able to get Charity Navigator wired up in the front end by tomorrow's 10am PT pre-demo, then.

**Comment by slifty** [comment]
>      I won't be able to get Charity Navigator wired up in the front end by tomorrow's 10am PT pre-demo, then.

There isn't much data to show anyway! Sorry if I should have spoken up when I said I wouldn't be working over the weekend.

File: src/candid.ts [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        const token = await getToken(
+          args.oidcBaseUrl as string,
+          args.oidcClientId as string,
+          args.oidcClientSecret as string,
+        );
+        const data = await getCandidProfile(args.candidApiKey as string, args.ein as string);
+        await postPlatformProviderData(
+          args.pdcApiBaseUrl as string,
+          token,
+          args.ein as string,
```
Comment (slifty) [inline src/candid.ts:115]: I just wanted to call out this pattern since I just got bit by it -- this violates type safety, since the actual value is not a string.  This means when doing something like `JSON.stringify(arg.sein)` the result is a number, not a string.

I don't quite know what magic yargs is implicitly doing around arguments (I can't tell where `ein` is actually defined, for instance), but it would be really nice to have all arguments be explicitly typed and provided to us as an explicitly typed value instead of shoehorned to pass type safety checks

File: src/candid.ts (hunk 5 above)
Comment (slifty) [reply src/candid.ts:109]: I'm not sure I understand the problem space, but does this help? https://github.com/yargs/yargs/blob/main/docs/typescript.md#typescript-usage-examples

File: src/candid.ts (hunk 9 above)
Comment (jasonaowen) [reply src/candid.ts:115]: Oh, that's great feedback! It sounds like I misunderstood how that works.

> I can't tell where `ein` is actually defined

It is defined as an argument to the `update` command via the angle brackets in the command definition:

https://github.com/PhilanthropyDataCommons/data-scripts/blob/9780b998c65f7f07ab87d5fbaf41affc9cb5773a/src/candid.ts#L102

See the docs for [positional arguments](https://github.com/yargs/yargs/blob/main/docs/advanced.md#positional-arguments).

Looks like there's a way to [specify the type of positional arguments](https://github.com/yargs/yargs/blob/main/docs/advanced.md#describing-positional-arguments), which I'd missed - sorry for leaving a trap that bit you, @slifty!

**Comment by jasonaowen** [comment]
In addition to rebasing to resolve merge conflicts, I'm also going to split out the infrastructure bits from the Candid-specific bits. That'll be in a separate PR, which this will ladder off.

File: README.md (hunk 1 above)
Comment (bickelj) [reply README.md:20]: Perhaps I made it sound like I personally was confused but my point is that npm makes it confusing. Of course I can look it up and dig around or try to memorize these distinctions. Thanks for taking the time to explain, though.

File: src/client.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
+import axios, { InternalAxiosRequestConfig, AxiosResponse } from 'axios';
```
Comment (bickelj) [inline src/client.ts:1]: It seems weird to have to use `InternalAxiosRequestConfig` which I assume was intended to be internal (despite being exported) rather than `AxiosRequestConfig` but the latter doesn't seem to work. :shrug:

Comment (bickelj) [reply src/client.ts:1]: I see [the issue was already reported](https://github.com/axios/axios/issues/5573). https://traveling-coderman.net/code/node-architecture/logging-http-requests/ shows using `AxiosRequestConfig` which must have worked in the past.

File: README.md (hunk 1 above)
Comment (bickelj) [reply README.md:20]: I was able to run `npm start -- auth ...` with the appropriate args and get a token. Nice.

Comment (bickelj) [reply README.md:20]: [yargs says](https://github.com/yargs/yargs#description) it provides bash completion, but when I type `npm start -- auth --[tab]` it's only looking for files. Not that I need it, but I was trying to see what goodies came out of the box with yargs.

File: src/pdc-api.ts [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      headers: {
+        authorization: `Bearer ${token.access_token}`,
+      },
+      data,
+    },
+  );
+  return response.data;
+};
+
+interface ApiBaseField {
```
Comment (bickelj) [inline src/pdc-api.ts:27]: I would prefer to omit the `Api`. Alternatives could be `BaseField` or `PdcBaseField` or `PDCBaseField`. I wouldn't even mention it except that this code appears intended to be common code for multiple scripts in future.

File: src/pdc-api.ts (hunk 4 above)
Comment (bickelj) [reply src/pdc-api.ts:43]: It would still be nice to have a single definition even within the `service` project where either the API spec gets generated from the original interfaces or vice versa. But I digress.

File: README.md (hunk 1 above)
Comment (jasonaowen) [reply README.md:20]: [Completion](https://github.com/yargs/yargs/blob/e517318cea0087b813f5de414b3cdec7b70efe33/docs/api.md#completioncmd-description-fn) is an opt-in feature, but that's worth playing with - thanks for calling it out!

File: src/pdc-api.ts (hunk 11 above)
Comment (jasonaowen) [reply src/pdc-api.ts:27]: This relates to a discussion we started in https://github.com/PhilanthropyDataCommons/data-viewer/issues/108 !

In particular, we found it useful to transform the data structures returned by the API to other client-specific data structures, and ended up having a hard time coming up with names for those data structures. These definitions were copied from the data-viewer, but I think the same constraints may apply - although probably not as strongly. Additionally, I think there's some merit in being consistent between these two client repos.

Calling it just "BaseField" gives these data structures a priority that might not be appropriate, does not communicate their origin, and makes search (or search-and-replace) more difficult. The `PDC` prefix doesn't suggest to me a distinction between API-facing and internal-facing data structures.

What do you think with all that additional context, @bickelj?

File: src/pdc-api.ts [hunk 12]
```diff
@@ -0,0 +1,100 @@
+import { client } from './client';
+import type { AccessTokenSet } from './oidc';
+
+const callPdcApi = async <T>(
+  baseUrl: string,
+  path: string,
+  params: Record<string, string>,
+  token: AccessTokenSet,
```
[the remark this answers is not in the document]
Comment (jasonaowen) [reply src/pdc-api.ts:8]: I think where I'm leaning now is to pass around a client (an `AxiosInstance`, specifically) that includes both the base URL, the authentication header, common logging options, and whatever else.

(I mentioned this as a possibility [earlier](https://github.com/PhilanthropyDataCommons/data-scripts/pull/24#discussion_r1206113769)

File: package.json (hunk 3 above)
Comment (jasonaowen) [reply package.json:36]: I rewrote the Winston logging commit to use pino instead, and split it out into its own PR: #48.

File: src/pdc-api.ts (hunk 4 above)
Comment (jasonaowen) [reply src/pdc-api.ts:43]: Yeah, no matter what approach we take, I don't think it's going to get done this month or as part of this PR.

I opened https://github.com/PhilanthropyDataCommons/service/issues/408 to track this issue.

**Comment by jasonaowen** [comment]
I've split off #48 for adding logging. After that, I'll add the yargs framework in a followup PR. Then, I'll rebase this PR - which will only focus on Candid - onto those.

File: README.md (hunk 1 above)
Comment (jasonaowen) [reply README.md:20]: Okay, I've spent some time playing with completion, and I think it's not worth doing, for a few reasons.

First, enabling bash completion involves evaluating some yargs-generated shell script in your current shell. This is something we'd want to keep up to date, which might be achievable with [direnv](https://direnv.net/) or the like, but is a bit risky - I don't love the idea of blindly executing external code upon entering a directory. Alternatively, each developer can periodically run `npm start -- completion >> ~/.bashrc` to pick up recently added commands, which is more secure and more likely to go out of date.

The other reason is that it'll take some doing to make the above work correctly. The way we're running this script is via `npm run-script`, and that adds some headers like `> data-scripts@1.0.0 start`. Unfortunately, the output of the `--get-yargs-completions` command - automatically added by `completion()` - is expected to be valid tab completions, and `>` is not a valid command. Additionally, the length of what should be `argv[0]` - `npm start --` makes it more complicated.

Finally, completion seems focused on commands, and not on options, which makes it less useful.

I think it could be made to work, but I also don't think it's worth the time. Which is unfortunate, because this is potentially a neat feature!

If you want to play with it further:

```diff
diff --git a/src/index.ts b/src/index.ts
index 61861b1..8320cf6 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -12,6 +12,7 @@ const main = async (argv: string[]) => yargs(hideBin(argv))
   .usage('Usage: npm start -- <command> [options]')
   .version(false)
   .strictCommands(true)
+  .completion()
   .env('DS')
   .config('config')
   .command(
```
```sh
npm start -- completion
```
```sh
npx ts-node src/index.ts --get-yargs-completions "candid"
```

File: src/pdc-api.ts (hunk 11 above)
Comment (bickelj) [reply src/pdc-api.ts:27]: I think it's good enough. My other thoughts can go in https://github.com/PhilanthropyDataCommons/service/issues/408.

File: README.md (hunk 1 above)
Comment (bickelj) [reply README.md:20]: Thanks for looking into the completion and I agree if it's not safe or easy to enable, it's not worth it.

My unspoken thought was "what benefits do we get for the cost of the additional complexity associated with yargs?" I vaguely recall other benefits. Also if it is a tool that most Node.js developers use then it might make things easier for new developers.

File: src/candid.ts (hunk 9 above)
Comment (jasonaowen) [reply src/candid.ts:115]: > I just wanted to call out this pattern since I just got bit by it -- this violates type safety, since the actual value is not a string. This means when doing something like `JSON.stringify(args.ein)` the result is an encoded number, not a string.

Hm, I wish I had dug into this more deeply while @slifty was around more. I haven't been able to reproduce this.

Positional arguments seem to default to be parsed as `string`s; entering numbers on the command line - and, in particular, numeric expressions like EINs - show up as `string`s for me. Specifying that a positional argument or an option is a `'number'` does cause it to be parsed as such.

The one wrinkle is that if you set an option to be a `string`, and then pass in a `--config` file that has the value as a JSON number, it will get passed in as a number. The reverse is not true: setting the option type to `number`, and passing in a `--config` file with a string-encoded numeric value will be parsed as a number.

---

I also looked more deeply into the original concern I had about needing to add type casts on use, and I think I figured out a solution! We'll need to declare a type for the expected arguments, a la React, and the handler should infer that its `args` is of that type:

```ts
interface ExampleCommandArgs {
  aRequiredString: string;
  anOptionalNumber?: number;
}
const exampleCommand: CommandModule<{}, ExampleCommandArgs> = {
  command: 'example',
  builder: {
    'a-required-string': { type: 'string', demandOption: true },
    'an-optional-number': { type: 'number' },
  },
  handler: (args) => {
    console.log(`It is safe to use string methods ${args.aRequiredString.substring(0)}`);
    if (args.anOptionalNumber) {
      assert(args.anOptionalNumber + 0 !== args.anOptionalNumber + '0');
    }
  },
};
```

This is better than adding `as string` or whatever to every use, although it does not rise to the level of runtime type assertions. This addresses my concern, although I don't know if it addresses @slifty's.

**Comment by jasonaowen** [comment]
I split out #52 for adding yargs.

File: src/candid.ts [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+
+const updateCommand: CommandModule = {
+  command: 'update <ein>',
+  describe: 'Fetch information about an organization by its EIN, and upload to the PDC',
+  builder: (y) => (y
+    .option('candid-api-key', {
+      describe: 'Candid Premier API key; get from account management at https://dashboard.candid.org/',
+      demandOption: true,
+      type: 'string',
+    })
```
Comment (jasonaowen) [inline src/candid.ts:87]: I don't love repeating this stanza from `lookup`.

File: src/candid.ts [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+    .check(({ ein }) => isValidEin(ein))
+  ),
+  handler: async (args) => {
+    const token = await getToken(
+      args.oidcBaseUrl as string,
+      args.oidcClientId as string,
+      args.oidcClientSecret as string,
```
Comment (jasonaowen) [inline src/candid.ts:105]: I couldn't figure out a way to avoid these as being typed as `unknown`.

File: src/pdc-api.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (jasonaowen) [reply src/pdc-api.ts:8]: I didn't get to this, I'm afraid.

**Review by jasonaowen — COMMENTED** [review]
This is as done as it will be this month, I'm afraid. I'm not completely happy with where I ran out of time, but it works, it's a mostly-decent example of how to add commands to the unified script, and it passes the linter.

Sorry to drop this at the end of the day at the end of the week at the end of the month!

File: src/candid.ts [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+  message: string;
+  data: object;
+}
+
+const getCandidProfile = async (
+  apiKey: string,
+  ein: string,
+): Promise<CandidPremierResult> => {
+  logger.debug(`Looking up EIN ${ein} in Candid Premier API`);
+  const { data } = await client.get<CandidPremierResult>(
```
Comment (bickelj) [inline src/candid.ts:20]: When this returns a 401 with `npm start candid lookup 12-3456789 -- --candid-api-key='testing'`, I don't see the logger boilerplate. It's not a problem, really, but it differs from the behavior exemplified in `npm start error`.

File: src/candid.ts (hunk 13 above)
Comment (bickelj) [reply src/candid.ts:87]: :shrug:

File: src/candid.ts [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+  return einBaseField.id;
+};
+
+const getEinsFromPdc = async (baseUrl: string, token: AccessTokenSet) => {
+  const einBaseFieldId = await getEinBaseFieldId(baseUrl, token);
+  const proposals = await getProposals(baseUrl, token);
+
+  const eins = new Set(proposals.entries
+    .flatMap((proposal) => proposal.versions[0]?.fieldValues)
+    .filter(<T>(x: T | undefined): x is T => typeof x !== 'undefined')
```
Comment (bickelj) [inline src/candid.ts:142]: My eyes had to pause here for a bit but then it makes sense. The next filter will get no "undefined" fields.

File: src/candid.ts [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+          ein,
+          'candid',
+          data,
+        );
+        logger.debug(`[${i + 1}/${eins.length}] Wrote data for ${ein} to PDC`);
+      } catch (error: unknown) {
+        logger.error({ error }, `Error loading data for ${ein}`);
+      }
+
+      // Our Candid API subscription has a rate limit of 10 calls per
```
Comment (bickelj) [inline src/candid.ts:213]: :disappointed:

File: src/candid.ts [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+       *
+       * Note that this use case is specifically called out in the ESLint rule
+       * documentation:
+       *
+       *     loops may be used to prevent your code from sending an excessive
+       *     amount of requests in parallel. In such cases it makes sense to
+       *     use await within a loop and it is recommended to disable the rule
+       *     via a standard ESLint disable comment.
+       *
+       * https://eslint.org/docs/latest/rules/no-await-in-loop
```
Comment (bickelj) [inline src/candid.ts:193]: :heavy_check_mark:

File: src/client.ts [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+import { logger } from './logger';
+
+const clientLogger = logger.child({}, { msgPrefix: '🌐 ' });
```
Comment (bickelj) [inline src/client.ts:4]: Haha, what? I like it.

File: src/pdc-api.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (bickelj) [reply src/pdc-api.ts:8]: I'd still prefer not to have the token here be strictly an `AccessTokenSet` defined in `oidc.ts` but the token string. If I find myself writing code to call this function later in another script, I might change it.

**Review by bickelj — APPROVED** [review]
I couldn't actually call Candid (no API key) but it looks good and trust that it works!

File: src/client.ts (hunk 19 above)
Comment (jasonaowen) [reply src/client.ts:4]: Aha, great catch, @bickelj! It turns out error handling behaves differently between synchronously throwing an error and asynchronously returning a Promise that later rejects. I've added a commit to fix that!

File: src/candid.ts (hunk 16 above)
Comment (jasonaowen) [reply src/candid.ts:142]: Yep, exactly! I believe that should have already been the case, but type-theoretically that wasn't previously guaranteed, and now it's certain.

File: src/candid.ts (hunk 15 above)
Comment (jasonaowen) [reply src/candid.ts:20]: Aha, great catch, @bickelj! It turns out error handling behaves differently between synchronously throwing an error and asynchronously returning a Promise that later rejects. I've added a commit to fix that!

File: src/client.ts (hunk 19 above)
Comment (jasonaowen) [reply src/client.ts:4]: You should be able to see it with `LOG_LEVEL=debug`: `LOG_LEVEL=debug npm start -- candid lookup 12345678`.

**Comment by jasonaowen** [comment]
I rebased onto main and added a commit to address a concern @bickelj [raised](https://github.com/PhilanthropyDataCommons/data-scripts/pull/24#discussion_r1330731490). I'll leave this open for a day or two in case anyone wants to review that new change, and then I'll merge.

**Comment by jasonaowen** [comment]
Oh hey, this has been open for a long time! :sweat_smile: 

Let's get this merged!
